### PR TITLE
Add `repeated_where_clause_or_trait_bound` lint

### PR DESCRIPTION
**Pull request by aldhsu**
I thought I would try and scratch my own itch for #8674.

1. Is comparing the `Res` the correct way for ensuring we have the same trait?
2. Is there a way to get the spans for the bounds and clauses for suggestions?
I tried to use `GenericParam::bounds_span_for_suggestions` but it only gave me an empty span at the end of the spans.
I tried `WhereClause::span_for_predicates_or_empty_place` and it included the comma.
3. Is there a simpler way to get the trait names? I have used the spans of the traits because I didn't see a way to get it off the `Res` or `Def`.

changelog: Add ``[`repeated_where_clause_or_trait_bound`]`` lint.

### CODE REVIEW

**Comment by aldhsu** [comment]
Good catch @giraffate! This also uncovers a bug in the existing TRAIT_DUPLICATION_IN_BOUNDS lint.

I'll have a think about it because it is a recursive problem. Do you have any suggestions for doing an easy comparison?

**Comment by flip1995** [comment]
You should rebase your branch on top of the current master. Some things changed with how generics are lowered in rustc just a few days ago. You might have to revisit some parts of the code.

**Comment by aldhsu** [comment]
Thanks for the heads up @flip1995. I have pushed a fixup that addresses the changes.

I left it unsquashed so how the generic lowering works now could be discussed.

I wasn't able to fix an issue with supporting the unstable associated_type_bounds feature.
Could you help me understand how the new lowering works with nested trait bounds? ie.
```rust
#![feature(associated_type_bounds)]

fn flatten_twice_good<T>(iter: T) -> Flatten<Flatten<T>>
where
    T: Iterator<Item: IntoIterator<Item: IntoIterator>>,
{
    iter.flatten().flatten()
}
```
https://github.com/rust-lang/rust-clippy/pull/8703/commits/c9c31f3be6734978e13bb5f5827c2818ed46bc77 handled this because bounds could contain their nested bounds. I haven't been able to follow how WhereBoundPredicate can nest WhereBoundPredicates yet.

**Comment by flip1995** [comment]
I wouldn't worry too much about supporting this feature with this lint. I'd suggest to just leave a `FIXME` comment that this should be addressed once this feature gets stabilized, which might take quite some time.

**Comment by aldhsu** [comment]
Squashed and added the fixme.

There is one other outstanding issue I haven't been able to figure out.

This case should be detected.
https://github.com/rust-lang/rust-clippy/blob/6040fa64323ebfcd906e6f039b35b776a07a5486/tests/ui/repeated_where_clause_or_trait_bound.rs#L40-L42

Doing some debugging on the Generic, no predicates exist. This only happens for the self trait bound and not the where clause. I haven't dug any further than that.

**Comment by flip1995** [comment]
For your example you should find those bounds in the [`ItemKind::Trait`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/enum.ItemKind.html#variant.Trait). It's the tuple field with the `GenericBounds` type.

**Comment by aldhsu** [comment]
Thanks for the tip @flip1995. I got access to an `ItemKind::Trait` by implementing `check_item` of the LateLintPass for Traitbounds. Is this what you meant?

**Comment by flip1995** [comment]
Thanks! I'd like to delegate the remaining review of this lint to @dswij. @dswij if you already have too many assigned PRs feel free to bounce it back to me though.

**Review by dswij — COMMENTED** [review]
Many thanks for the PR! Changes overall looks great to me. It's a really nice addition for sure. 

Just left one nit that's not really important to the change.

---

> Is comparing the Res the correct way for ensuring we have the same trait?

Comparing Res within a type bound should be fine.

> Is there a simpler way to get the trait names? I have used the spans of the traits because I didn't see a way to get it off the Res or Def.

Well, we could've used `ident`. But for the suggestions here, using `snippet_opt` with the spans would be preferred

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs:367]: Very cool.

File: tests/ui/repeated_where_clause_or_trait_bound.rs
Comment (aldhsu) [reply tests/ui/repeated_where_clause_or_trait_bound.rs:6]: Added that test case in https://github.com/rust-lang/rust-clippy/pull/8703/commits/e6b34d7b3dfcb15608754ce5b55054bb09b32eab and one for a user specified with a similar name.

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs]: Renamed to repeated_res.

**Comment by aldhsu** [comment]
Thanks for the review @dswij. I have actioned it.

**Review by dswij — APPROVED** [review]
Thanks for addressing the comments! It looks good to me.

I'm not sure whether this lint should be in `nursery` group though, I don't see this having lots of false warnings either. Need some opinion on this, what do you guys think? @xFrednet @flip1995

**Comment by xFrednet** [comment]
Thank you for the review @flip1995 offered to do the final review (probably after their vacation :))

> I'm not sure whether this lint should be in nursery group though, I don't see this having lots of false warnings either. Need some opinion on this, what do you guys think?

I think the final group should be `pedantic`. `TRAIT_DUPLICATION_IN_BOUNDS` which is similar, is currently in the nursery group due to some braking changes in rustc. I believe that they're fixed now, but I'm not totally sure. If they are fixed, we should change the lint group. Until then, I would keep this in the same group as `TRAIT_DUPLICATION_IN_BOUNDS`. @flip1995 is more up-to-date on this AFAIK :upside_down_face:

**Comment by aldhsu** [comment]
@dswij I put it in the nursery group because the other trait bound ones were put in there. I didn't have the confidence to say that the new lint was ready when the original lints it was based off were moved to nursery 😅. Original group was `pedantic` as @xFrednet said. 

Offtopic: What is the process of growing out of nursery? I'm going to pick up https://github.com/rust-lang/rust-clippy/issues/8757

**Comment by xFrednet** [comment]
> Offtopic: What is the process of growing out of nursery? I'm going to pick up #8757

It depends on the reason why the lint was moved to the nursery group. The general rule is, that lints can be moved to their respective group once all issues have been resolved.

In this case I remember somebody working on fixing the issues. I'm just not sure if they've finished their work. It could be, that they've been fixed on the rust side and will be enabled again during the weekend. I'll try to keep an eye on it :)

File: clippy_lints/src/trait_bounds.rs
Comment (flip1995) [inline clippy_lints/src/trait_bounds.rs:97]: Could this lint be included in the `TRAIT_DUPLICATION_IN_BOUNDS`? When thinking about a good lint name for this lint I thought `TRAIT_DUPLICATION_IN_BOUNDS` would actually the best name for this lint.

If this lint should be kept its own separate lint, the lint name should be

```suggestion
    pub REPEATED_WHERE_CLAUSES_OR_TRAIT_BOUNDS,
```

File: clippy_lints/src/trait_bounds.rs
Comment (flip1995) [inline clippy_lints/src/trait_bounds.rs:98]: I fixed the issue in rustc and moved the other lints back to pedantic: https://github.com/rust-lang/rust/pull/96770

So regardless whether this lint is merged with the other lint or kept on its own, it should be in pedantic.

File: clippy_lints/src/trait_bounds.rs
Comment (flip1995) [inline clippy_lints/src/trait_bounds.rs:99]: This should be lower case, just like lint messages. I also noticed that this is also wrong for the other lints in this file. If you want you can also change this for those.
```suggestion
    "traits are repeated within trait bounds or where clause"
```

**Review by flip1995 — COMMENTED** [review]
Thanks for the review @dswij and thanks for addressing it @aldhsu!

I left some comments. Not sure if this requires a new lint.

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs]: Fixed in https://github.com/rust-lang/rust-clippy/pull/8703/commits/6e2ec70ce3a8ed0f15e33bafdcee400e22bf725c.

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs]: Moved it back to pedantic.

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs]: I think it is possible to merge the two. 

I have a proposal since merging `trait_duplication_in_bounds` and `repeated_where_clause_or_trait_bound` in could be too much work and might mean the PR stalls. 

I propose we:
1. Merge this PR when it is acceptable without adding any more functionality.
2. Fix https://github.com/rust-lang/rust-clippy/issues/8757 and merge PR. I've been working on a fix that is based on some work in this PR to make comparable trait refs. This will also make the two more similar and easier to refactor.
3. I'll work on a refactor PR to merge `trait_duplication_in_bounds` and `repeated_where_clause_or_trait_bound`.

**Comment by aldhsu** [comment]
Thanks for the review @flip1995!

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs]: @flip1995 what do you think of this proposal?

File: clippy_lints/src/trait_bounds.rs
Comment (flip1995) [inline clippy_lints/src/trait_bounds.rs:381]: Merging the lints would just mean to change this line and adapt the documentation of the other lint a bit. I don't really want to merge a new lint, that will be renamed later, because if we mistime things, the temporary lint name will get into stable and we have to go through the renaming procedure or backport things.

**Review by flip1995 — COMMENTED** [review]
So sorry, I completely forgot about this PR... I would do the renaming now, it shouldn't be too much work. If you lost the motivation to work on this after that long time, I'll address this myself in order not to lose this awesome work.

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs:381]: @flip1995 ah got it. Of course the lint would be just under the same name! I was thinking it had to be done in the same pass.

**Comment by aldhsu** [comment]
@flip1995 I can address it. I had it in my head that "merging" the lint meant doing it in the same pass. I should be able to get it done this weekend.

Thanks for calling it awesome!

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs:381]: Merging under the existing lint name raises some extra noise with UI test errors since the original lint "catches" some extra errors (#8757). Since this is actually a bug, I have left a comment that the errors generated in `stderr` should be removed once #8757 is fixed. 

With regards to fixing, I've just turned off rustfix since I can't get the fixed file to compile without fixing the original lint.

Once this PR is merged I can work on reducing the noise in the errors so that it's not reporting the same errors twice slightly differently. This sounds ok to me since the lint is back in nursery.

File: tests/ui/trait_duplication_in_bounds.rs
Comment (flip1995) [inline tests/ui/trait_duplication_in_bounds.rs:197]: If this is the only test case that trips up rustfix, I'd suggest to move it into its own file named something like `tests/ui/trait_duplication_in_bounds_8757.rs`. That way, you can keep rustfix enabled and still have a regression test for the issue.

Or are there other cases that would break rustfix?

Comment (aldhsu) [reply tests/ui/trait_duplication_in_bounds.rs:197]: Yes, there are other cases that break rustfix.

The existing lint of `trait_duplication_in_bounds` raises a lint for pretty much all the new examples added. I believe the original intention of the lint is supposed to only be for traits appearing in both trait bounds and where clauses as that seems to be the only kinds of tests. I think it accidentally picks up duplicates **within** trait bounds and where clauses. This is evident because the errors produced for these cases are misleading.

Example:
```rust
    fn bad_foo<T: Clone + Clone + Clone + Copy, U: Clone + Copy>(arg0: T, argo1: U) {
        unimplemented!();
    }

```

```
error: this trait bound is already specified in the where clause
  --> $DIR/trait_duplication_in_bounds.rs:103:19
   |
LL |     fn bad_foo<T: Clone + Clone + Clone + Copy, U: Clone + Copy>(arg0: T, argo1: U) {
   |                   ^^^^^
   |
   = help: consider removing this trait bound

```
`this trait bound is already specified in the where clause`
It's error refers to a where clause but there aren't any. 

I can fix this issue along with #8757 .
At that time I can split the fixables and unfixables like other I see done with the other test files.

Comment (flip1995) [reply tests/ui/trait_duplication_in_bounds.rs:197]: Ah ok, that's interesting. It seems that the old lint already triggers on the cases, the newly added lint code should check for, just with a worse error message + suggestion. I think this is even worse than #8757. Since this lint is in `nursery` I'd be ok merging this. But I would open an issue about it, even when you start working on resolving this right away.

File: clippy_lints/src/trait_bounds.rs
Comment (aldhsu) [reply clippy_lints/src/trait_bounds.rs]: Resolved in https://github.com/rust-lang/rust-clippy/pull/8703/commits/a0c2da924ea381c221bcbdd06436a1c9c96936e1

File: tests/ui/trait_duplication_in_bounds.rs
Comment (aldhsu) [reply tests/ui/trait_duplication_in_bounds.rs:197]: Issue here #9151

**Comment by aldhsu** [comment]
@flip1995 squashed. Thanks for the review!

**Comment by flip1995** [comment]
Thanks! Sorry again for forgetting about this PR

@bors r+

**Pull request merged**
